src/Microsoft.AspNet.Hosting/HostingEngine.cs

```diff
@@ -39,16 +40,16 @@ public IDisposable Start(HostingContext context)
             InitalizeServerFactory(context);
             EnsureApplicationDelegate(context);
 
-            var applicationLifetime = (ApplicationLifetime)context.Services.GetRequiredService<IApplicationLifetime>();
+            var applicationLifetime = (ApplicationLifetime)context.ApplicationLifeTime;
```

Comment:
This cast makes it impossible to replace this service. Maybe the interface needs to be re-worked.